[syntax-errors] `return` outside function

ntBre opened the pull request:
Summary
--

This PR reimplements [return-outside-function (F706)](https://docs.astral.sh/ruff/rules/return-outside-function/) as a semantic syntax error.

These changes are very similar to those in https://github.com/astral-sh/ruff/pull/17298.

Test Plan
--

New linter tests, plus existing F706 tests.

dhruvmanila commented on crates/ruff_python_parser/src/parser/statement.rs:
```diff
@@ -254,7 +254,8 @@ impl<'src> Parser<'src> {
 
         // test_ok simple_stmts_with_semicolons
-        // return; import a; from x import y; z; type T = int
+        // def outer():
```
nit: I think the function names don't matter here so would prefer to just keep it `_`

dhruvmanila commented on crates/ruff_linter/src/checkers/ast/mod.rs:
```diff
@@ -586,6 +588,11 @@ impl SemanticSyntaxContext for Checker<'_> {
                     ));
                 }
             }
+            SemanticSyntaxErrorKind::ReturnOutsideFunction => {
+                if self.settings.rules.enabled(Rule::ReturnOutsideFunction) {
+                    self.report_diagnostic(Diagnostic::new(ReturnOutsideFunction, error.range));
+                }
+            }
```
What is the plan for this and other related semantic syntax errors that are implemented as rules? Should this be converted as other semantic syntax errors and deprecate the rules when stabilizing these syntax errors?

ntBre replied:
I think the long-term plan is to deprecate these as rules, but I'm not sure if stabilization will happen too soon for that deprecation to be included. I'm interested in other thoughts on that!

The main benefit of reimplementing them like this is that red-knot will be able to reuse the syntax error versions, regardless of how they continue to be used in ruff.

But in short, yes, this whole function should eventually just be `self.semantic_errors.borrow_mut().push(error);`.

dhruvmanila replied:
Yeah, that makes sense, thanks.

dhruvmanila commented on crates/ruff_python_parser/src/semantic_errors.rs:
```diff
@@ -118,12 +118,22 @@ impl SemanticSyntaxChecker {
+                    // def f(): return *x
+                    Self::invalid_star_expression(value, ctx);
+                }
+                if !self.in_function_scope {
```
As stated in the other PR, I'd prefer if we can re-use the scope information from the `Checker`.

ntBre commented:
Could I get one more quick review here? Like in https://github.com/astral-sh/ruff/pull/17298, I reverted the parser snapshot changes and moved everything to a linter test. It's much, much smaller now.

The function scope tracking is now on the `Context` too, of course.
